### PR TITLE
Allow backing up etcd cluster

### LINKED ISSUE #1318: Allow backupPolicy to be set on self-hosted etcd clusters
We'd like to back up our etcd state with a [BackupPolicy](https://github.com/coreos/etcd-operator/blob/master/pkg/spec/backup.go#L32) but the current manifest is [hard-coded](https://github.com/coreos/tectonic-installer/blob/master/modules/bootkube/resources/experimental/etcd/migrate-etcd-cluster.json#L8-L35). What would be the best way to parameterize this? 

I'm thinking we could maybe change the `template_dir` resource defined [here](https://github.com/coreos/tectonic-installer/blob/master/modules/bootkube/assets.tf#L24) to a series of `template_file` resources (one per file), and then load either `migrate-etcd-cluster.json` or `migrate-etcd-cluster-with-backup.json` based on a new bool var.

### PR DESCRIPTION
**Pull request by jamiehannaford**
Fixes #1318 

/cc @s-urbaniak

### CODE REVIEW

**Comment by s-urbaniak** [comment]
ok to test

**Comment by s-urbaniak** [comment]
@jamiehannaford thanks for the PR! CI complains about the following:
```
In ../modules/bootkube/resources/bootkube.sh line 16:
  if [ -x ${enable_etcd_backup} ] || [ ${enable_etcd_backup} != true ]; then
          ^-- SC2154: enable_etcd_backup is referenced but not assigned.
          ^-- SC2086: Double quote to prevent globbing and word splitting.
                                       ^-- SC2086: Double quote to prevent globbing and word splitting.
```

**Comment by jamiehannaford** [comment]
@s-urbaniak I'm not sure why `SC2154: enable_etcd_backup is referenced but not assigned.` is appearing, since that var is injected in via [this line](https://github.com/coreos/tectonic-installer/pull/1831/files#diff-86c16ffa7f01163f45b284b61ee3c662R171) right?

**Comment by s-urbaniak** [comment]
@jamiehannaford I think you're right, and the linter confuses bash vars with terraform variables. I suggest adding `# shellcheck disable=SC2154` like [here](https://github.com/coreos/tectonic-installer/pull/1831/files#diff-09b320f560f88a35a49662a7603525c8R27).

**Comment by jamiehannaford** [comment]
@s-urbaniak I added it before the conditional (see [here](https://github.com/coreos/tectonic-installer/pull/1831/commits/785b990ce5ef1fcdb26901735a88dca6df9afc76#diff-09b320f560f88a35a49662a7603525c8R16)) but the CI is still failing. Is it the same error?

**Comment by cpanato** [comment]
@jamiehannaford this is the error in the CI

```

In ../modules/bootkube/resources/bootkube.sh line 17:
  if [ -x ${enable_etcd_backup} ] || [ ${enable_etcd_backup} != true ]; then
          ^-- SC2086: Double quote to prevent globbing and word splitting.
                                       ^-- SC2086: Double quote to prevent globbing and word splitting.

```

**Comment by cpanato** [comment]
@jamiehannaford 
CI now complains:

```
Error loading modules: module bootkube: Error loading .terraform/modules/0391b25f8df845582a6b6a555d42296b/variables.tf: Variable 'enable_etcd_backup' type must be one of [string, map, list] - 'boolean' is not a valid type
Makefile:36: recipe for target 'terraform-get' failed
```

**Comment by jamiehannaford** [comment]
@cpanato Hmm, I thought `boolean` was a valid type 🤔  What is the latest error?

**Comment by cpanato** [comment]
@jamiehannaford 
```
* module root: 
    module bootkube: required variable "authentication_token_webhook_url" not set
    module bootkube: required variable "etcd_backup_storage_class" not set

```

**Comment by jamiehannaford** [comment]
@cpanato @s-urbaniak What is the latest error?

**Comment by cpanato** [comment]
@jamiehannaford was an internal Jenkins error. I will retrigger the job.

**Comment by cpanato** [comment]
ok to test

**Comment by cpanato** [comment]
@jamiehannaford error now is in the fmt, can you please fix when you have some time?
```
misformatted files (run 'terraform fmt .' to fix): config.tf modules/update-payload/config.tf platforms/aws/config.tf platforms/azure/config.tf platforms/metal/config.tf platforms/openstack/neutron/config.tf platforms/vmware/config.tf
```

**Comment by jamiehannaford** [comment]
@cpanato Okay, I've done that. What does the CI say? We'll get there in the end 😅

**Comment by cpanato** [comment]
@jamiehannaford still in the lint part, try `run 'make docs' to fix`. 

we are together in this :)

````
terraform-docs --no-required markdown  config.tf    >>  Documentation/variables/config.md
terraform-docs --no-required markdown  platforms/aws/variables.tf    >>  Documentation/variables/aws.md
terraform-docs --no-required markdown  platforms/azure/variables.tf    >>  Documentation/variables/azure.md
terraform-docs --no-required markdown  platforms/openstack/neutron/variables.tf    >>  Documentation/variables/openstack-neutron.md
terraform-docs --no-required markdown  platforms/metal/variables.tf    >>  Documentation/variables/metal.md
terraform-docs --no-required markdown  platforms/vmware/variables.tf    >>  Documentation/variables/vmware.md
make[1]: Leaving directory '/home/ubuntu/workspace/tectonic-installer_PR-1831-CGZBDC2V7WR5HXYFPXAA6O6IZUTWOWBONBT4KLM2GWKHCQJHAHIQ'
diff --git a/Documentation/variables/config.md b/Documentation/variables/config.md
index 87891bab..b09eab29 100644
--- a/Documentation/variables/config.md
+++ b/Documentation/variables/config.md
@@ -22,6 +22,9 @@ This document gives an overview of variables used in all platforms of the Tecton
 | tectonic_ddns_key_name | (optional) This only applies if you use the modules/dns/ddns module.<br><br>Specifies the RFC2136 Dynamic DNS server key name. | string | `` |
 | tectonic_ddns_key_secret | (optional) This only applies if you use the modules/dns/ddns module.<br><br>Specifies the RFC2136 Dynamic DNS server key secret. | string | `` |
 | tectonic_ddns_server | (optional) This only applies if you use the modules/dns/ddns module.<br><br>Specifies the RFC2136 Dynamic DNS server IP/host to register IP addresses to. | string | `` |
+| tectonic_enable_etcd_backup | (Optional) Indicates whether self-hosted etcd clusters should be backup up to a persistent volume.<br><br>If enabled, a StorageClass must be specified using the tectonic_etcd_backup_storage_class variable. | string | `false` |
+| tectonic_etcd_backup_size | (optional) The size of the PersistentVolume used for handling etcd backups | string | `512` |
+| tectonic_etcd_backup_storage_class | (optional) The name of an existing Kubernetes StorageClass that will be used for handling etcd backups | string | - |
 | tectonic_etcd_ca_cert_path | (optional) The path of the file containing the CA certificate for TLS communication with etcd.<br><br>Note: This works only when used in conjunction with an external etcd cluster. If set, the variables `tectonic_etcd_servers`, `tectonic_etcd_client_cert_path`, and `tectonic_etcd_client_key_path` must also be set. | string | `/dev/null` |
 | tectonic_etcd_client_cert_path | (optional) The path of the file containing the client certificate for TLS communication with etcd.<br><br>Note: This works only when used in conjunction with an external etcd cluster. If set, the variables `tectonic_etcd_servers`, `tectonic_etcd_ca_cert_path`, and `tectonic_etcd_client_key_path` must also be set. | string | `/dev/null` |
 | tectonic_etcd_client_key_path | (optional) The path of the file containing the client key for TLS communication with etcd.<br><br>Note: This works only when used in conjunction with an external etcd cluster. If set, the variables `tectonic_etcd_servers`, `tectonic_etcd_ca_cert_path`, and `tectonic_etcd_client_cert_path` must also be set. | string | `/dev/null` |
outdated docs (run 'make docs' to fix)
```

**Comment by jamiehannaford** [comment]
@cpanato Done

**Comment by squat** [comment]
@jamiehannaford looks like there is one more linting error. You need to run `make examples`. This should be the last linting problem! In general, whenever the config.tf file or the platform specific variables.tf files change we need to:
```
terraform fmt
make docs
make examples
```

**Comment by jamiehannaford** [comment]
@squat @cpanato Is there anything left to do?

**Comment by cpanato** [comment]
@jamiehannaford saw that
```

cd /jenkins/workspace/tectonic-installer_PR-1831-CGZBDC2V7WR5HXYFPXAA6O6IZUTWOWBONBT4KLM2GWKHCQJHAHIQ/build/tf-metal-PR-1831-16 && TERRAFORM_CONFIG=/jenkins/workspace/tectonic-installer_PR-1831-CGZBDC2V7WR5HXYFPXAA6O6IZUTWOWBONBT4KLM2GWKHCQJHAHIQ/.terraformrc terraform apply  /jenkins/workspace/tectonic-installer_PR-1831-CGZBDC2V7WR5HXYFPXAA6O6IZUTWOWBONBT4KLM2GWKHCQJHAHIQ/platforms/metal
var.tectonic_etcd_backup_storage_class
  (optional) The name of an existing Kubernetes StorageClass that will be used for handling etcd backups

  Enter a value: 
var.tectonic_etcd_backup_storage_class
  (optional) The name of an existing Kubernetes StorageClass that will be used for handling etcd backups

  Enter a value: 
var.tectonic_etcd_backup_storage_class
  (optional) The name of an existing Kubernetes StorageClass that will be used for handling etcd backups

  Enter a value: 
var.tectonic_etcd_backup_storage_class
  (optional) The name of an existing Kubernetes StorageClass that will be used for handling etcd backups

  Enter a value: 
Error asking for user input: missing required value for "tectonic_etcd_backup_storage_class"
Makefile:44: recipe for target 'apply' failed
```

**Comment by jamiehannaford** [comment]
@cpanato Hmm. Do I need to populate `tectonic_etcd_backup_storage_class` with a dummy value in all the tests/smoke/*/vars files?

**Comment by cpanato** [comment]
my understanding if those are optional if should not ask if you not set it.

@alexsomesan or @s-urbaniak can give a hand here, please?

File: config.tf [hunk 1]
```diff
@@ -408,3 +408,26 @@ WARNING: Enabling an alpha feature means that future updates may become unsuppor
...
+
+variable "tectonic_etcd_backup_size" {
+  type        = "string"
+  description = "(optional) The size of the PersistentVolume used for handling etcd backups"
+  default     = "512"
+}
+
+variable "tectonic_etcd_backup_storage_class" {
+  type        = "string"
+  description = "(optional) The name of an existing Kubernetes StorageClass that will be used for handling etcd backups"
```
Comment (s-urbaniak) [inline config.tf]: if marked as `(optional)`, we must provide a default value here.

File: config.tf [hunk 2]
```diff
@@ -408,3 +408,26 @@ WARNING: Enabling an alpha feature means that future updates may become unsuppor
 This should only be enabled on clusters that are meant to be short-lived to begin validating the alpha feature.
 EOF
 }
+
+variable "tectonic_enable_etcd_backup" {
+  type = "string"
+
+  description = <<EOF
+(Optional) Indicates whether self-hosted etcd clusters should be backup up to a persistent volume.
```
Comment (s-urbaniak) [inline config.tf]: s/`(Optional)`/`(optional)` (lowercase)

**Comment by s-urbaniak** [comment]
@jamiehannaford left a few comments, why the optional logic fails in this PR, PTAL.

File: config.tf (hunk 1 above)
Comment (jamiehannaford) [reply config.tf]: AFAIK there is no default StorageClass in Kubernetes, so does that mean this value needs to be required? If so, I'm concerned that users who _don't_ want to back up etcd will need to provide a value, even though they have no interest in using it. 

Maybe we could use a default placeholder value like "default"?

Comment (s-urbaniak) [reply config.tf]: In this case I suggest to add `default = ""` and make it clear in the description that this value must be set iff `tectonic_enable_etcd_backup` is set to `true`. Unfortunately terraform does not give us more expressiveness in terms of inter-variable dependencies.

**Comment by s-urbaniak** [comment]
Also /cc @sym3tri and @robszumski since this is 1. adding another manifest 2. adding quite a new feature product-wise.

File: config.tf (hunk 2 above)
Comment (s-urbaniak) [reply config.tf]: small nit: Do you mind to add `(optional) [ALPHA] Indiciates whether ...`?

**Comment by jamiehannaford** [comment]
@s-urbaniak I added "" as the value and regenerated all the docs. What is left to do?

**Comment by cpanato** [comment]
@jamiehannaford I will re-run the job again I think was some flakes. I will confirm that bear with me

**Comment by cpanato** [comment]
@jamiehannaford tests passed, except for one flake azure test.

@s-urbaniak its ready for your review

**Comment by s-urbaniak** [comment]
ack, ok I'd love to get this in, but let me sync with @sym3tri, whether this in can live in-tree. We are adding a new manifest here, which potentially will have to be supported by KVO. Having said that, if we declare this as experimental (as we do), that's probably (?) fine.

**Comment by jamiehannaford** [comment]
@s-urbaniak Any update on this? Also, I'm not sure why the build is suddenly failing. Any ideas?

**Comment by s-urbaniak** [comment]
@jamiehannaford sorry for the delay, since this affects core functionality I'd like to have a green light roadmap-wise first, ping @sym3tri.

(also needs a rebase)

**Comment by jamiehannaford** [comment]
@s-urbaniak @sym3tri Okay. Hopefully this doesn't pose too much of a change since it's optional and feature gated by the experimental flag.

**Comment by jamiehannaford** [comment]
@s-urbaniak @cpanato What is the build failure?

**Comment by cpanato** [comment]
@jamiehannaford re running that, something bad on jenkins happened :/

**Comment by cpanato** [comment]
@jamiehannaford @s-urbaniak saw this in the CI
```

Error downloading modules: Error loading modules: module bootkube: Error parsing .terraform/modules/0391b25f8df845582a6b6a555d42296b/variables.tf: At 152:3: invalid characters in heredoc anchor
Makefile:38: recipe for target 'terraform-init' failed
```

maybe a rebase fix that.

**Comment by jamiehannaford** [comment]
@cpanato What's the failure for?

**Comment by jamiehannaford** [comment]
Ping :) Would quite like to get this in soon

**Comment by squat** [comment]
@jamiehannaford the smoke tests are complaining that `config.tf` is improperly formatted. Please run `terraform fmt config.tf` to fix it.

**Comment by jamiehannaford** [comment]
I've done that, still errors. Is it possible to expose Jenkins so folks can see why their builds failed?

**Comment by lblackstone** [comment]
@sym3tri Looks like this is still waiting on your approval. Could you give a 👍  / 👎  on this feature? Thanks!

**Comment by sym3tri** [comment]
@s-urbaniak I'm ok with this once @robszumski signs off. I pinged him.

**Comment by robszumski** [comment]
LGTM @sym3tri

**Comment by jamiehannaford** [comment]
Thanks @robszumski!

@squat @cpanato Any idea what the latest CI error is? All syntax, docs and examples seem to be correct.

**Comment by s-urbaniak** [comment]
@jamiehannaford all looks good, thanks a lot for the patience! I rekicked the build as the failure is a flake.

**Comment by jamiehannaford** [comment]
@s-urbaniak Ping, is this still a CI flake?

**Comment by squat** [comment]
ok to test

**Comment by squat** [comment]
Looks like some azure tests flaked

File: platforms/azure/tectonic.tf [hunk 3]
```diff
@@ -89,7 +89,11 @@ module "bootkube" {
 
   master_count = "${var.tectonic_master_count}"
 
-  cloud_config_path = "/etc/kubernetes/cloud"
+  cloud_config_path = ""
```
Comment (squat) [inline platforms/azure/tectonic.tf]: Why is the cloud provider configuration path being unset here?

**Comment by jamiehannaford** [comment]
@squat Good catch, that was a typo. Sorry for that. Looks like the build is now passing 🎉

**Comment by s-urbaniak** [comment]
ok to test

**Comment by s-urbaniak** [comment]
@jamiehannaford thanks a lot for the patience! I rekicked the built, as the appropriate labels have not been set.

**Comment by jamiehannaford** [comment]
@s-urbaniak CI might need kicking again

**Comment by s-urbaniak** [comment]
@jamiehannaford sorry to reping, do you mind to rebase on master, as GH doesn't seem to sync this up well. azure had a ci failure which we just fixed on master.

**Comment by jamiehannaford** [comment]
@s-urbaniak Rebased, CI still needs a kick

**Comment by s-urbaniak** [comment]
@jamiehannaford CI rekicked, but I this time we have a legit failure, this is the output of `tectonic.service
```
xxxx-master-0 bash[773]: [  287.440693] bootkube[5]: All self-hosted control plane components successfully started
xxxx-master-0 bash[773]: [  287.440856] bootkube[5]: Migrating to self-hosted etcd cluster...
xxxx-master-0 bash[773]: [  287.441053] bootkube[5]: I1025 13:55:35.463041       5 log.go:19] warning: ignoring ServerName for user-provided CA for backwards compatibility is deprecated
xxxx-master-0 bash[773]: [  307.115019] bootkube[5]: I1025 13:55:55.137957       5 migrate.go:65] created etcd cluster CRD
xxxx-master-0 bash[773]: [  312.315920] bootkube[5]: I1025 13:56:00.338862       5 migrate.go:76] etcd-service IP is 10.3.0.15
xxxx-master-0 bash[773]: [  317.389236] bootkube[5]: Error: failed to create etcd cluster for migration: open /assets/etcd/migrate-etcd-cluster.json: no such file or directory
xxxx-master-0 bash[773]: [  317.389527] bootkube[5]: Tearing down temporary bootstrap control plane...
xxxx-master-0 bash[773]: [  317.389678] bootkube[5]: Error: failed to create etcd cluster for migration: open /assets/etcd/migrate-etcd-cluster.json: no such file or directory
xxxx-master-0 bash[773]: [  317.389835] bootkube[5]: Error: failed to create etcd cluster for migration: open /assets/etcd/migrate-etcd-cluster.json: no such file or directory
xxxx-master-0 bash[773]: [  317.389975] bootkube[5]: failed to create etcd cluster for migration: open /assets/etcd/migrate-etcd-cluster.json: no such file or directory
```

**Comment by s-urbaniak** [comment]
it seems the moved `migrate-etcd-cluster.json` must be reflected in the azure platform

**Comment by jamiehannaford** [comment]
@s-urbaniak I renamed the file, but still no idea if it's passed CI :) Would you mind outputting the logs either here or on Slack?

**Comment by s-urbaniak** [comment]
I'll debug it here on my side, stay tuned :-)

**Comment by s-urbaniak** [comment]
@jamiehannaford please allow me introduce a superseded PR for this. I would suggest not to introduce yet another `etcd_backup_enabled` variable, next to `tectonic_experimental`.

`tectonic_experimental` has no meaning other than self hosted etcd and we wanted to get rid of that variable anyway.

Instead I factored locally a solution which introduces one variable `tectonic_self_hosted_etcd` replacing `tectonic_experimental` and `tectonic_enable_etcd_backup` which is a string and can be set to the following values:

- `""` (empty string): this means we deploy a "standard" dedicated etcd cluster, this is the default
- `"enabled"`: this deploys a self-hosted etcd
- `"pv_backup"`: this deploys a self-hosted etcd including pv based backups

sorry for holding off on this, but it has been on my plate anyways.

I will ammend to your commit in the superseded PR to retain your original authorship :-)

**Comment by jamiehannaford** [comment]
That's awesome, thanks @s-urbaniak! Feel free to close this whenever.

**Comment by s-urbaniak** [comment]
closing in favor of #2232

**Pull request closed without merging**